### ISSUE TITLE
All properties in body component exposed and editable in graphs

### ISSUE THREAD

**Issue by TobiasElemental** [OurMachinery/themachinery-public#1113]
Atm only kinematic check-box is editable through scripting.
I would like to also:

- disable gravity
- set linear and angular axis locks
- set linear and angular damping

**Comment by Tisten** [comment]
The implementation of the kinematic graph node is today located in the physx plugin, while I would like to have all such graph nodes in the physics plugin instead. The reason is that we right now have a separate physx backend from the default, and are also looking at a havok backend. It is much easier to handle all changes to the component in the plugin which has the component, and then just write the code which transfers the properties to the backend in the backend.

Another issue with the kinematic graph node implementation is that it applies the propery immediately to the backend while the simulation is running, instead of just setting it on the `tm_physics_body_component_t` and waiting with applying it until the backend can accept it. This works with PhysX 4.1 since it is double buffered during simulation, but neither havok or other versions of PhysX has this double buffering.

If you agree with my stance on this then I can start moving the kinematic graph node and create a PR for that, plus adding nodes for the properties mentioned in this issue.

If you want to keep it as it is then I will wait for your implementation for this issue, and then write new implementations for the new backend(s).

**Comment by niklas-ourmachinery** [comment]
Seems reasonable to have these nodes live in `physics`. 

Though we may at some point get into sticky situation when with different backends... i.e. what happens if a particular backend doesn't support "locks" or "disable gravity".

The double buffering approach seems reasonable to... in fact it would be good if it was possible to modify `physics` component properties and have them mirrored to `PhysX` (or other backends).

One thing we might need in the future is to have the ability to run engines only over changed components... i.e. instead of looping over all the body components, we loop only over those for which the flag has changed. We're currently thinking about what is the best way of doing those kind of things.

**Comment by Tisten** [comment]
@leonardo-ourmachinery reached out to me about my ideas and pointed me to his big refatoring of all physics mirroring https://github.com/OurMachinery/themachinery/pull/879. I think any work on this issue should build on top of his PR, and I appreciate that Leo read my comments and is considering how his changes and my ideas could merge in the best way.

The problem of different backends having different feature sets is a problem with all abstraction layers. Either you need to add emulation layers on top so all backends can perform the same thing, or expose the least common denominator (providing you know all the backends), or you have a query mechanism in the backend, and the frontend will only expose what the backend can do, but that makes it harder to ship the same game with two different backends on different platforms. I know you have a lot of experience and knowledge about creating good APIs so I wont rant more about this here.

Right now we are using two different backends, one which is stable and the other one which we are hardening, i.e it is still in development. So right now it is important that they expose the same properties and behave the same for us.

Having knowledge about if components are active or resting would be a good property to build optimizations on. What this state means is component specific, and it can depend on other components in a hierarchy or other relationship. But adding utilities to utilize such a property if it exists, that would be a great thing, and it absolutely applies to physics components. I guess it also ties into the discussion about disabling/enabling things in https://github.com/OurMachinery/themachinery-public/issues/1111

**Comment by leonardo-ourmachinery** [comment]
I think that we should also consider potentially just changing the _names_ of the graph nodes but leave them where they are. (inside the Physx plugin).

This means that the Physx plugin will expose nodes that will be called something like `physics_set_kinematic` and `physics_is_kinematic` instead of `physx_set_kinematic` and `physx_is_kinematic`.

This way:
-The Physx plugin can implement the nodes however it wants. In this case everything could remains as it is as the Physx version we're using has double buffering.
-Another physics backend could implement the same nodes and still expose them with the same `physics_set_kinematic` and `physics_is_kinematic` names, but queue the changes in its own internal queue.
-If a backend doesn't support a specific feature it just won't implement the corresponding graph node. This way if a backend has gravity and another doesn't and you switch from the former to the latter, you will receive a message like "cannot find `physics_has_gravity` node" or something like that.

The big advantage of having the nodes live under the physics plugin is that you're writing the graph code only once. (Setting the value in the component data)... but then you will still have to write the "mirroring" code for every different backend.

**Comment by Tisten** [comment]
I wouldn't like changing the names of the nodes and leaving them in the backend, because then switching between different backends might change which inputs and outputs the nodes have.

Also, since the types are defined in the physics plugin, the type will still have the "gravity" property even if the backend doesn't have the node. So to me physics should either define all or nothing, i.e either no components and no graph nodes, or define both. And to avoid divergence between plugins my preference is towards physics defining both, and then writing the mirroring code with static asserts so if the number of properties change when you integrate from the public master then you get a compile error.

**Comment by leonardo-ourmachinery** [comment]
Yeah, you are right. We would need to have the "declaration" of the node to live in the physics plugin, and then each specific backend _could_ provide the implementation of the node.
It's not impossible, but it doesn't sound something that's worth implementing. Even if I think it would be pretty in line with the PR about the "virtual" physx components.

Regarding the "gravity" property, I could argue that some properties may not even be stored in the component because the backend is already keeping a copy of it internally anyway... but yes we're entering a pretty dangerous ground here, so I agree that the best first step is to move the nodes to the physics plugin and put the property "mirroring" mechanism in place.

**Comment by leonardo-ourmachinery** [comment]
@niklas-ourmachinery @Tisten I just realized we have the Entity Property system that could be used to solve this problem quite nicely in my opinion: the `set_xxx_property` could flag the component as dirty automatically in addition to modifying the component data, and then the physics backend could loop just on the dirty components and mirror stuff.

Basically with this approach we would just get rid of the Physx graph nodes altogether and there wouldn't be the need for writing other nodes manually in the Physics plugin, but just expose entity properties from the components.

The only real drawback I can see is that only properties supported by the Entity Property system would work out of the box, but I think that would be fine, and we can always think about supporting more in the future.

**Comment by niklas-ourmachinery** [comment]
I think that would be a good approach... and the long term goal is to have most properties exposed in this system anyway.

I think a drawback is that it's a little more indirect to have a generic "Set Property" node where you have to type in the name of the property instead of a node that just does exactly the thing you want ("Set Mass"... etc) But I'm not sure how big of a problem it is in practice.

**Comment by Tisten** [comment]
I took a stab at this to get to know the Entity Property system.

I tried it for this issue, i.e for the body component.

I felt that in the "set" callback function, the `tm_entity_property_i` data is missing. This means that each property needs its own callback function, instead of having a generic function.

Another issue is that in the body component there are currently 9 bools (3 bools and one bitfield with 6 used bits), and property system only have float types.

And as Niklas pointed out, using the "Set Float Property" node type user experience is worse than using dedicated nodes, especially since the property name always is too big to fit in the node, so you can't understand what the node does without having written it in the label, or clicking on the node and reading the properties tab.

So in the short term I would favor writing specific nodes.
And the improvements I would like to see before converting it to the Entity Property system are:
* Add support for all primitive types to the system
* Feed the `tm_entity_property_i` data to get and set callbacks
* Add support for nodes to generate a label if no label is explicitly set

**Comment by niklas-ourmachinery** [comment]
Not sure if I see the value of the generic `set()` function... a custom `set()` function could just call the generic `set()` if so desired.

The property system deliberately abstracts all data as `float`s so that it can be treated in a uniform way (for curve animation, etc). A bool is abstracted as a float with values `0` and `1`. I think there is a value to this abstraction, so I am reluctant to introduce additional types.

**Comment by Tisten** [comment]
I respect those design descisions, that this system is made for systems which want to pay for an abstraction since they need a uniform interface for access. But for Graph Nodes it is clearer with explicit connector types, so based on that design I don't think the Entity Property system is an advantage compared to writing nodes which are dedicated for the components. The result will be worse ui and ux, more code (copy paste style, repetitive, deeper callstacks to debug), more expensive to execute and reduced functionality.

The only potential advantage I can think of is unrelated to Graph Nodes: If there is an advantage of hiding component structs in the private data, so the only way to access (write to) the members is via the property system. But that would eliminate any cheap access to them in ECS code outside of the plugin.

**Comment by Tisten** [comment]
In that case I can make a PR for this issue for you since I added nodes for
this during last week.

There are a few things I haven't done, one is how to treat the relationship
of "mass", "center of mass" and the "inertia" properties. They are
connected, but I just treat them as individual properties. Secondly I kept
the physx name for them even though I moved them to physics. I ought to
write my first migration code instead, but then I should also move move
nodes than just the ones handling the body properties.
